text/0000-system-buildpacks-in-builder.md

```diff
@@ -0,0 +1,155 @@
+# Meta
+[meta]: #meta
+- Name: System Buildpacks in Builder Images
+- Start Date: 2021-07-24
+- Author(s): [@jkutner](https://github.com/jkutner)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+This is a proposal for a mechanism that would allow a builder to contain a default set of buildpacks that run on every build, independent of which buildpack group passes detection.
+
+# Definitions
+[definitions]: #definitions
+
+* _system buildpacks_ - a standard buildpack, conforming to the Buildpack API, but which runs outside of normal groups
+
+# Motivation
+[motivation]: #motivation
+
+Forthcoming changes to the lifecycle (such as [removal of shell-specific logic](https://github.com/buildpacks/rfcs/pull/168)) will remove capabilities that users have come to expect. This includes mechanisms like `.profile`, which allows a buildpack user to customize the environment a process type runs in. We seek to replace these lost mechanisms with buildpacks, in an effort to preserve the capability while still removing complexity from the lifecycle.
+
+# What it is
+[what-it-is]: #what-it-is
+
+We introduce a `[system]` table in the `builder.toml` schema with the following structure:
+
+```
+[[system.pre.buildpacks]]
+  id = "<buildpack ID>"
+  version = "<buildpack version>"
+  optional = false
+
+[[system.post.buildpacks]]
+  id = "<buildpack ID>"
+  version = "<buildpack version>"
+  optional = false
+```
+
+The fields in the `system.pre.buildpacks` table and `system.post.buildpacks` table match the fields in the existing [`order.group` table](https://buildpacks.io/docs/reference/config/builder-config/#order-_list-required_).
+
+When a builder includes one or more `system.*.buildpacks` entry, each build phase that uses the builder will run the `pre` buildpacks before ALL other buildpacks (whether they are defined by a default `order.group` or if they are provided as an explicit order to the lifecycle) and the `post` buildpack after ALL other buildpacks.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+Unless otherwise stated, system buildpacks conform to the [buildpack API](https://github.com/buildpacks/spec/blob/main/buildpack.md).
+
+The `system.pre.buildpacks` will be transformed into a new table in `order.toml`. The `[system]` table in `order.toml` will be processed by the lifecycle, and each `pre`/`post` buildpack will run during the detect phase. Those that pass detection will run during the build phase.
+
+## Detection
+
+The results of detection by system buildpacks MUST NOT influence the selected buildpack group. If no system buildpacks pass detection, any buildpack group MAY pass detection. If a system buildpack pass detection and no buildpack groups pass detection, then detection MUST fail.
```

Comment:
> The results of detection by system buildpacks MUST NOT influence the selected buildpack group.

I'm not sure I understand this requirement. If system buildpacks can augment the build plan, they will also influence the selected buildpack group. 

> If no system buildpacks pass detection, any buildpack group MAY pass detection.

Similar issue to the above -- certain groups may require a system buildpack to include build plan entries.

> If a system buildpack pass detection and no buildpack groups pass detection, then detection MUST fail.

Is this just to prevent groups with all optional buildpacks from turning into groups that only contain system buildpacks?

Reply:
>>  The results of detection by system buildpacks MUST NOT influence the selected buildpack group.

> I'm not sure I understand this requirement. If system buildpacks can augment the build plan, they will also influence the selected buildpack group.

I really meant for this to be about exit code. With regard to the build plan, maybe the system buildpack should have either not have that, or have a separate build plan.

Reply:
I think that I would like to add a +1 to @sclevine point. I don't understand why these wouldn't just be treated as buildpacks that you want to run before and after every other group. Maybe I am missing something bigger here but I feel like the concept to me was simpler if it is essentially just matrix expansion if you will. And with the ability to mark the buildpacks optional I think you still have a very wide range of flexibility.

Reply:
I'm hoping to avoid having our decision to decouple certain behaviors (like `.profile`) from the implementation of the lifecycle manifest itself in the user experience. 

I think buildpack _users_ shouldn't need to understand or care how things like "optional" or groups, or any other buildpack concepts work. They should run `pack build` (or whatever for their platform) and it just detects a python app, and builds it. Adding buildpacks like `.profile` or whatever into the build breaks the abstraction provided by `pack`. (also maybe this decision can be scoped to pack and not lifecycle).

Ultimately though, I think this comes down to _how much_ we decide to push into the system/utility buildpacks. If it's really just `.profile` in most cases, then I probably don't care as much. But I worry about a situation where it used to be as simple as "the `heroku/python` buildpack passed detect" to "why are there 8 buildpacks running?" Again I think that comes back to a difference b/w Heroku/Paketo buildpack design.

Reply:
Does this mean the system buildpack can't write to the build plan and only participates with the exit code? I think that'd make sense in the `.profile` case.